### DISCUSSION TITLE
[mini.diff] compare with staged version?

### DISCUSSION THREAD

**Discussion by hosewiejacke** [nvim-mini/mini.nvim#1333] [Q&A]
Currently, mini.diff compares files with their committed version.

Is it possible to compare files with their staged version?

Background:
Sometimes I skim over the changes (`]h`) and stage everything I'd like to commit (`gh`). As of now, there is no indication in the sign column, whether a hunk is staged or not. Would it be possible to have different signs or colors for modified/staged hunks?

Btw: thanks for mini.diff. I switched to it from gitsigns :-)

**Comment by echasnovski** [comment]
I think there is some misunderstanding here.

'mini.diff' has a more general model than working only with Git: it has buffer (i.e. current) text and reference text. Two of them are compared with each other whenever each of them changes. How reference text is computed depends on the buffer source. You can read more [here](https://github.com/echasnovski/mini.nvim/blob/7b4d5d48b6b5a75009d63f8f3e4ef4819b7e8139/doc/mini-diff.txt#L83).

The default source [compares buffer's text with the text in the Git index](https://github.com/echasnovski/mini.nvim/blob/7b4d5d48b6b5a75009d63f8f3e4ef4819b7e8139/doc/mini-diff.txt#L20-L21), i.e. the text in the Git staging area, i.e. staged version of the file. Here are a bit [more details](https://github.com/echasnovski/mini.nvim/blob/7b4d5d48b6b5a75009d63f8f3e4ef4819b7e8139/doc/mini-diff.txt#L528-L536).

> Sometimes I skim over the changes (`]h`) and stage everything I'd like to commit (`gh`). As of now, there is no indication in the sign column, whether a hunk is staged or not. Would it be possible to have different signs or colors for modified/staged hunks?

Staging hunk means adding it to the index. This means that the text version in the buffer and in the reference text are the same. As there is no difference, there is no diff hunk to act (show, navigate, etc.) upon. Hence nothing is shown (in sign/number column, in overlay, etc.)

------

Does this clarify things about how 'mini.diff' works?

**Comment by hosewiejacke** [comment]
Thanks  a lot! Yes, it did :-)

After some more investigation this seems to be a UI glitch in my environment. E.g. after applying a hunk (via `gh`) the UI isn't updated:

![image](https://github.com/user-attachments/assets/f0289c02-806c-4074-91b2-f05cfd64a3c7)

I first need to reload the file with `:e` to see the applied hunk:
 ![image](https://github.com/user-attachments/assets/028df6ea-e6e2-444f-983d-c35c397ad27a)

**Reply by echasnovski** [reply]
> After some more investigation this seems to be a UI glitch in my environment. E.g. after applying a hunk (via `gh`) the UI isn't updated:

This should not happen. Here is how it works for me locally (with 'signcolumn=auto'):

https://github.com/user-attachments/assets/7512cd49-e330-413c-9589-332c0a2c1afa

What Neovim version and OS are you using? And just to be sure, did you disable 'gitsigns.nvim' (so that it doesn't run in parallel with 'mini.diff')?

**Comment by hosewiejacke** [comment]
Thx, I really appreciate you taking the time to look at this.

- yes, I disabled gitsigns.nvim
- signcolumn=auto, too
- Neovim version:  v0.10.1 (LuaJIT 2.1.1713484068)
- OS: OpenBSD 7.6, amd64

To be honest, Gitsigns has the same problem (not updating the UI after applying a hunk). The problem there is luv:
https://github.com/lewis6991/gitsigns.nvim/blob/main/lua/gitsigns/watcher.lua#L146-L152

Maybe this is OpenBSD related. I don't want to bother you with stuff you're not using....

**Reply by echasnovski** [reply]
Yeah, it *might* be the same reason. Staging hunk in `gen_source.git()` indeed relies on directory/file watching to update the reference text. Sorry I can't help any further here.
